### PR TITLE
update ap_sta_config.sh

### PR DESCRIPTION
**Pull request by ipitio**
* Added some changes from a pull request to rpi-wifi: https://github.com/lukicdarkoo/rpi-wifi/pull/26
* Added 802.11n support to hostapd.conf
* Fixed bash variable syntax and a typo in manage-ap0-iface.sh
* Disabled wifi controller power_save
* Added path to commands in rpi-wifi.sh

### CODE REVIEW

**Comment by MkLHX** [comment]
Hi @ipitio 

Thanks for contribution. Sorry for late reply i was never notified by github about this.
Could you please ask a little bit more about what is the goal of it's PR.

I ask for that because this script is used every day in production in our devices and i need to be sure before merge anything.

**Review by MkLHX — COMMENTED** [review]
why use systemctl status hostapd instead of service??

**Review by MkLHX — COMMENTED** [review]
why add /sbin/ before each command?
There is a power save mode on Wifi on buster? I never see anything about this.
It's turned on by default?

**Comment by ipitio** [comment]
Hi, sorry for not elaborating! The changes from rpi-wifi#26 may actually be outdated as per [the latest commit](https://github.com/lukicdarkoo/rpi-wifi/pull/26/commits/58ee4d8e32a82ffbb3a790a92064caf3e297d44d). In hostapd.conf, the change to enable 802.11n and wmm may result in better speeds and the change to the wpa parameters fix syntax.

> why use systemctl status hostapd instead of service??

`service` is a wrapper for `systemctl` so there isn't really a reason to use one over the other. 

> why add /sbin/ before each command?

If someone doesn't have `PATH` set in the proper place, the current script will return "Command not found" errors. Prepending the full path (in this case `/sbin`) to each command [will prevent this](https://unix.stackexchange.com/a/163126).

> There is a power save mode on Wifi on buster? I never see anything about this.
> It's turned on by default?

There is a power save mode, though I'm not sure whether it's turned on by default; you can check whether it's enabled for wlan0 with `iw wlan0 get power_save`. If it is ever enabled for wlan0 or ap0, it [may result in an unreliable connection](https://www.raspberrypi.org/forums/viewtopic.php?t=292697), so it's best to make sure it's off.

**Comment by MkLHX** [comment]
Thx for reply @ipitio 

I need to improve changes on my side. Have you push these modifications on you repo? Can i use it for testing the changes?

**Comment by ipitio** [comment]
Just pushed the latest changes from #26, didn't test yet. The source given for those changes includes instructions for configuring a bridge between eth0 and ap0, which I can implement later.

**Comment by MkLHX** [comment]
@ipitio 
Ok so i can't merge for now i have to test it before but haven't the time for now too sorry.

**Comment by wohaoa** [comment]
@ipitio I executed your branch, but got the following error:
`ap_sta_config.sh: line 244:DAEMON_OPTS: unbound variable`

File: ap_sta_config.sh [hunk 1]
```diff
@@ -189,178 +188,115 @@ if test true != "${STA_ONLY}" && test true == "${AP_ONLY}"; then
...
 
-if test true != "${STA_ONLY}"; then
     # Populate `/etc/hostapd/hostapd.conf`
     _logger "Populate /etc/hostapd/hostapd.conf"
-    bash -c 'cat > /etc/hostapd/hostapd.conf' <<EOF
+    bash -c 'test -f /etc/hostapd/hostapd.conf || cat > /etc/hostapd/hostapd.conf' << EOF
 ctrl_interface=/var/run/hostapd
 ctrl_interface_group=0
-interface=ap0
+interface=ap0@wlan0
```
Comment (davidmurdoch) [inline ap_sta_config.sh:227]: should this be `ap@wlan0`?

File: ap_sta_config.sh [hunk 2]
```diff
@@ -189,178 +188,115 @@ if test true != "${STA_ONLY}" && test true == "${AP_ONLY}"; then
...
 EOF
-fi
 
-if test true != "${STA_ONLY}"; then
     # Populate `/etc/dnsmasq.conf`
     _logger "Populate /etc/dnsmasq.conf"
-    bash -c 'cat > /etc/dnsmasq.conf' <<EOF
-interface=lo,ap0
+    bash -c 'cat > /etc/dnsmasq.conf' << EOF
+interface=lo,ap0@wlan0
```
Comment (davidmurdoch) [inline ap_sta_config.sh:212]: should this be `ap@wlan0`?

**Comment by akumpf** [comment]
> @ipitio I executed your branch, but got the following error: `ap_sta_config.sh: line 244:DAEMON_OPTS: unbound variable`

This error seems to be caused by a leftover variable declaration from the previous version. I replaced the variables that were causing the error, and it seems to be working well for me so far:
* Search for `$DAEMON_OPTS ${DAEMON_CONF}` and replace with `/etc/hostapd/hostapd.conf` 

edit: I also made the changes @davidmurdoch suggested (`ap0@wlan0` changed to `ap@wlan0`, which occurs twice).

File: ap_sta_config.sh (hunk 1 above)
Comment (ipitio) [reply ap_sta_config.sh:227]: It could be, but convention is to name it apX

**Comment by ipitio** [comment]
Thanks @akumpf! What did you replace the variables with?

@MkLHX @davidmurdoch 
I've also come across [RaspAP](https://github.com/RaspAP/raspap-webgui) and [linux-wifi-hotspot](https://github.com/lakinduakash/linux-wifi-hotspot), which may interest you.

**Comment by akumpf** [comment]
@ipitio I updated my previous comment to be more explicit.

@MkLHX -- is there any doc/explanation around the `ap_sta_cron.sh` script? I see it's commented out of this update. Is that something that just needs to be run/installed once? Or re-run after changes are made to the AP/STA parameters?

**Comment by MkLHX** [comment]
> Thanks @akumpf! What did you replace the variables with?
> 
> @MkLHX @davidmurdoch I've also come across [RaspAP](https://github.com/RaspAP/raspap-webgui) and [linux-wifi-hotspot](https://github.com/lakinduakash/linux-wifi-hotspot), which may interest you.

  Thank you. I know RasAP but it to more complex for my use case but it's a really good project.
  I don't know linux-wifi-hotspot i'll go to see that in depth.

**Comment by MkLHX** [comment]
> @ipitio I updated my previous comment to be more explicit.
> 
> @MkLHX -- is there any doc/explanation around the `ap_sta_cron.sh` script? I see it's commented out of this update. Is that something that just needs to be run/installed once? Or re-run after changes are made to the AP/STA parameters?

cron job manager is script launch at the en of ap_sta_config.sh
Stange behavior on raspberry pi doesn't allow me to edit crontab using ap_sta_config.sh
so i split it i n 2 files and call the last at the end
[here](https://github.com/MkLHX/AP_STA_RPI_SAME_WIFI_CHIP/blob/master/ap_sta_config.sh#L392)

**Comment by MkLHX** [comment]
Also i have to repeat my post on september 2020 here:
https://github.com/MkLHX/AP_STA_RPI_SAME_WIFI_CHIP/pull/3#issuecomment-918928508
i can't merge for now, i have to take the time to reveiw every edit from every peoples.
Here this master version is used on @greenponik product so it's really touchy for me to merge withtout testing the edits and be sure the feature still work on our products.
Also it miss some comments about why edition is submit so it's to hard to say "OK " i merge it.

**Comment by akumpf** [comment]
@MkLHX @ipitio @davidmurdoch I also spotted a couple other things (like a reference to `/bin/rpi-wifi.sh`, which this update no longer populates)... but also fixable.

What _is_ super nice about this version:
* it leverages services (i.e. no cron) for monitoring.
* it doesn't hard-code a MAC address (i.e. you can clone the filesystem, and then boot on another pi, though haven't tested that myself yet).
* It's all contained in one script (no secondary cron helper script needed); simpler to debug and tweak.
* it seems to work well on the latest Raspberry Pi OS distro (Bullseye) and can work on the new Pi Zero 2 W. The other/main version breaks the wifi and it becomes unreachable on the Pi Zero 2 W; not sure what the cause is, but the fact that this new version worked was a nice surprise.

Best approach might be to keep the existing scripts as-is, and create a new V2.x.x script (named `ap_st_config2.sh`?) that uses this updated approach -- I'll create a new pull with an updated version as the new file. Then perhaps it can be merged in as a secondary/experimental script until it has more eyes on it.

**Comment by MkLHX** [comment]
@ipitio i merged https://github.com/MkLHX/AP_STA_RPI_SAME_WIFI_CHIP/pull/12
on master so i think i can close this PR. What do you think?

**Comment by ipitio** [comment]
Yep!

**Pull request closed without merging**